rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/SW360RestHealthIndicator.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Bosch.IO GmbH 2020
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.sw360.rest.resourceserver;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import org.apache.thrift.TException;
+import org.eclipse.sw360.datahandler.common.DatabaseSettings;
+import org.eclipse.sw360.datahandler.couchdb.DatabaseInstance;
+import org.eclipse.sw360.datahandler.thrift.ThriftClients;
+import org.eclipse.sw360.datahandler.thrift.health.HealthService;
+import org.eclipse.sw360.datahandler.thrift.health.Status;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.stereotype.Component;
+
+import java.net.MalformedURLException;
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+public class SW360RestHealthIndicator implements HealthIndicator {
+    @Override
+    public Health health() {
+        List<Exception> exceptions = new ArrayList<>();
+        RestState restState = check(exceptions);
+        final String rest_state_detail = "Rest State";
+        if (!restState.isUp()) {
+            Health.Builder builderWithDetails = Health.down()
+                    .withDetail(rest_state_detail, restState);
+            for (Exception exception : exceptions) {
+                builderWithDetails = builderWithDetails.withException(exception);
+            }
+            return builderWithDetails
+                    .build();
+        }
+        return Health.up()
+                .withDetail(rest_state_detail, restState)
+                .build();
+    }
+
+    private RestState check(List<Exception> exception) {
+        RestState restState = new RestState();
+        try {
+            restState.isDbReachable = isDbReachable(exception);
+        } catch (MalformedURLException e) {
+            restState.isDbReachable = false;
+            exception.add(e);
+        }
+        restState.isThriftReachable = isThriftReachable(exception);
+        return restState;
+    }
+
+    private boolean isDbReachable(List<Exception> exception) throws MalformedURLException {
+        DatabaseInstance databaseInstance = makeDatabaseInstance();
+        try {
+            return databaseInstance.checkIfDbExists(DatabaseSettings.COUCH_DB_ATTACHMENTS);
+        } catch (Exception e) {
+            exception.add(e);
+            return false;
+        }
+    }
+
+    private boolean isThriftReachable(List<Exception> exception) {
+        HealthService.Iface healthClient = makeHealthClient();
```

Comment:
I am doing it the way it is done in the rest of the code, except that I extracted the call `new ThriftClients().makeHealthClient()` into a separate function.

Would you like me to not do it in a separate method?

Reply:
The code is fine.
I was trying to say, since thrift client wont change, maybe we can create a single `HealthService.Iface` object and reuse it.
Instead of creating a new one with every request.

```
    private static HealthService.Iface healthService;
    protected HealthService.Iface makeHealthClient() {
        if(healthService==null) {
            healthService = new ThriftClients().makeHealthClient();
        }
        return healthService;
    }
```

Reply:
I was under the impression that this is not possible due to Thrift not being threadsafe.
Since the calls are made via Spring we can't control the threads. Hence the client could be corrupted.
Since I am not sure what exactly thrift does under the hood to its client in a thread I would opt for the safest option, which is to declare a new one for each thread.